controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -163,37 +163,58 @@ public void initializeStreamWriters(final ClientFactory clientFactory,
+     * @param stream     stream name.
+     * @param streamCut  stream cut.
+     * @param contextOpt optional context
+     * @return update status.
+     */
+    public CompletableFuture<UpdateStreamStatus.Status> truncateStream(final String scope, final String stream,
+                                                                       final Map<Integer, Long> streamCut,
+                                                                       final OperationContext contextOpt) {
+        final OperationContext context = contextOpt == null ? streamMetadataStore.createContext(scope, stream) : contextOpt;
 
-        return streamMetadataStore.getState(scope, stream, true, context, executor)
-                .thenCompose(state -> streamMetadataStore.getConfigurationWithVersion(scope, stream, context, executor)
-                        .thenApply(configWithVersion -> {
-                            if (configWithVersion.getVersion() > prevVersion) {
-                                if (configWithVersion.getConfiguration().equals(update)) {
-                                    return !state.equals(State.UPDATING);
-                                } else {
-                                    throw new RuntimeException("Failed to update. Conflict.");
-                                }
-                            } else if (!state.equals(State.UPDATING)) {
-                                throw new RuntimeException("Failed to update. Conflict.");
-                            } else {
-                                return false;
-                            }
-                        }));
+        // 1. get stream cut
+        return streamMetadataStore.getTruncationProperty(scope, stream, true, context, executor)
+                .thenCompose(property -> {
+                    if (!property.isUpdating()) {
+                        // 2. post event with new stream cut if no truncation is ongoing
+                        return writeEvent(new TruncateStreamEvent(scope, stream))
+                                // 3. start truncation by updating the metadata
+                                .thenCompose(x -> streamMetadataStore.startTruncation(scope, stream, streamCut,
+                                        context, executor))
+                                // 4. check for truncation to complete
+                                .thenCompose(truncation -> checkDone(() -> isTruncated(scope, stream, streamCut, context))
+                                        .thenApply(y -> UpdateStreamStatus.Status.SUCCESS));
+                    } else {
+                        log.warn("Another truncation in progress for {}/{}", scope, stream);
+                        return CompletableFuture.completedFuture(UpdateStreamStatus.Status.FAILURE);
+                    }
+                })
+                .exceptionally(ex -> {
+                    log.warn("Exception thrown in trying to update stream configuration {}", ex.getMessage());
```

Comment:
you can just pass ex and leave out the {} that way we will get the stack trace

Reply:
done